Show active preset in lib module header

zisoft opened the pull request:
This PR adds the currently active preset to the header of lib modules, in a similar way as it is done in the iop modules:

![Bildschirmfoto 2025-02-23 um 18 33 58](https://github.com/user-attachments/assets/a7e37757-d1f0-4cbb-9236-aed3453870f4)

The preset label is named `lib-module-name` (analogous to the iop module `iop-module-name`) so it can be styled separately in css.

closes #18430 
closes #17442 
closes #16964
 closes #19020

zisoft committed:
add module preset label to module header

zisoft committed:
update preset_label on param change

zisoft committed:
rename static functions with leading '_'

dterrahe commented on src/libs/filtering.c:
```diff
@@ -2252,6 +2250,12 @@ void gui_init(dt_lib_module_t *self)
 }
 
+void gui_update(dt_lib_module_t *self)
+{
```
Either make gui_update a DEFAULT member function in lib_api (and provide a default_ implementation) or test if module->gui_update exists before calling it. No need for empty dummy implementations everywhere.

zisoft replied:
Great, thanks!

dterrahe replied:
That was just after a quick look at the code. I can't test or fully analyze at the moment, but beware, the reason for gui_update for libs is different from gui_update for iops. The former is meant to avoid (possibly) costly updates of the _content_ of a module if it is collapsed or otherwise not visible. For these changes, presumably you need to update the header even if the module is collapsed. If that means you are now always calling gui_update for each signal received, that may cause a more sluggish experience.

Again, I haven't analyzed your code.

zisoft committed:
make gui_update a default function

TurboGit reviewed (changes requested):
Still some fixes needed.
- The preset label on metadata & collection filter are not kept, select a preset it is displayed and then cleared just after **if the module is expanded**.

Note that in the console I get a huge amount of:
`(darktable:364759): Gtk-CRITICAL **: 21:48:12.549: gtk_label_set_text: assertion 'GTK_IS_LABEL (label)' failed`

And at some point I get this message continuously on the console.

The image information and export modules are ok.

zisoft committed:
fix gtk critical errors

TurboGit reviewed (changes requested):
Thanks, the gtk-critical messages are fixed. Two remaining issue:

- The collection filter module doesn't keep the preset label when it is expanded, select for example Square. If collapse it keeps the label, but as soon as you expand the module the label is cleared.
- Likewise for metadata editor module

zisoft commented:
@TurboGit: I am still working on this but I must admit that I'm thinking about giving up. The metadata module is the killer here, no other module refreshes its whole gui when hovering over the lighttable images. Updating the preset label in the module header works fine if the module is expanded. If the module is collapsed the gui data is not updated.

This is all well designed code (why should the gui data be updated if the gui is not visible?). Updating the preset label requires to compare the module params with the presets. All expensive operations. For now I have not found a proper way to get the params for a collapsed module.

However, I have fixed a bug in the metadata module which can be reproduced in master:

- reset the module
- apply a preset (i.e. "all rights reserved"): The fields get filled.
- open the preset menu and see the selected preset did not get a checkmark.

I have fixed that with commit 2ff0dbabfe5813dc81a3d1538f0733060c29d599 which can be cherry-picked for master if we drop this PR.

zisoft commented:
> The collection filter module doesn't keep the preset label

We have a general problem with the collection filter presets, tested in master:
The preset "initial setting" is the only one which gets marked with a checkmark. None of the other presets get that checkmark.

I'll make some analysis on that.

elstoc commented:
Can we have this either use the existing preference (darkroom > automatically update module name) to disable or implement a new one for lib modules

zisoft committed:
make module preset label opional

TurboGit commented:
@zisoft : Still on my radar, I hope to look at this before the end of the week.

TurboGit reviewed (commented):
Just a comment, this works for me and it seems all good.

Just that the name is always set and not based on the preference. At the moment the preference to name a module automatically is on the Darkroom's section. What about moving it to "Miscellaneous / Interface" and using it also for the 2 lighttable modules export/import? I would say that this will be more consistent. What your thoughts?

zisoft commented:
> I would say that this will be more consistent. What your thoughts?

Agreed, will do.

zisoft committed:
use preference setting for lib module preset name

zisoft commented:
@TurboGit : I have now moved that preference to miscellaneous/interface:

<img width="644" height="374" alt="Bildschirmfoto 2026-05-04 um 09 28 18" src="https://github.com/user-attachments/assets/251e1a1c-d2ba-49ba-9894-95e523f93061" />

If unset, the lib module preset label is not displayed.
I am unsure if the preference text is understandable in that way?

TurboGit commented:
> I am unsure if the preference text is understandable in that way?

Looks ok to me especially with the corresponding tooltip. I don't have a better proposal.

TurboGit commented on data/darktableconfig.xml.in:
```diff
@@ -168,6 +168,13 @@
     <shortdescription>condensed panels' controls</shortdescription>
     <longdescription>use condensed panels' controls in all views</longdescription>
   </dtconfig>
+  <dtconfig prefs="misc" section="interface" common="yes" restart="true">
```
We don't *really* need a restart. When you activate the option the preset name is added immediately. When you disable the option, the name will be removed when reset is called, not a big deal to me.

zisoft replied:
Ok, I had to change that a bit more. The name was removed on reset, but not on preset change. That should work now as well.

zisoft commented:
Release notes: The active preset name is now displayed in the header of suitable library modules (import, export). This can be turned off in the "miscellaneous" section in the preferences ("automatically update module name").

TurboGit reviewed (approved):
Thanks!

TurboGit commented:
@zisoft : You may want to look at this possible issues (me playing again with OpenCode :)

### 1. `free()` vs `g_free()` allocator mismatch in `_free_module_info` (MEDIUM)

`src/libs/lib.c:466`: Changed from `g_free(minfo->params)` to `free(minfo->params)`. The `params` pointer comes from `module->get_params()`, which may return memory allocated with `g_malloc` (varies by module). On platforms where glib uses a different heap (e.g., Windows debug builds), this is UB. The old code used `g_free`, which is safe for both `malloc` and `g_malloc` on most glib platforms.

**Fix**: Revert to `g_free(minfo->params)`.

### 2. `dt_lib_gui_update` fires expensive SQL query on every draw (MEDIUM)

`src/libs/lib.c:921-928`: The preset label update code runs **outside** the `gui_uptodate` guard, so it executes on every single draw event for every expandable module. For each draw:
- `_get_module_info_for_module` calls `get_params` (allocates/copies)
- `dt_lib_get_active_preset_name` compiles a SQL query + iterates rows + `memcmp`s blobs
- This runs even for modules with `has_preset_label()` returning `FALSE` (the widget exists, so `module->preset_label` is non-NULL)

With N expandable modules visible, that's N SQL queries per draw frame.

**Fix**: Guard with `module->has_preset_label(module)` check before the expensive block, and/or cache the active preset name and only re-query when params actually change.

### 3. No NULL guard in `_lib_gui_reset_callback` and `dt_lib_reset_preset_label` (LOW)

`src/libs/lib.c:971`: Calls `gtk_label_set_text(GTK_LABEL(module->preset_label), "")` without NULL-checking `module->preset_label`.

`src/libs/lib.c:1681`: Same in `dt_lib_reset_preset_label` — no NULL check on `mod` or `mod->preset_label`.

`_lib_gui_reset_callback` is called from `_action_process` (keyboard shortcut) for **any** module with `gui_reset`, but `preset_label` is only created in `dt_lib_gui_get_expander` for expandable modules. Currently no module has `gui_reset` without an expander, but this is a latent crash.

**Fix**: Add `if(!module->preset_label) return;` at the top of `_lib_gui_reset_callback`. Add `if(!mod || !mod->preset_label) return;` in `dt_lib_reset_preset_label`.

### 4. Redundant/non-standard `#include <glib-2.0/glib.h>` (LOW)

`src/libs/lib.c:33`: Uses `<glib-2.0/glib.h>` which relies on `/usr/include` being in the default compiler search path. The conventional path is `<glib.h>` (resolved via pkg-config `-isystem` paths). Additionally, glib.h is already transitively included through `lib.h` -> `<gtk/gtk.h>`, through `lib_api.h`, and through `module_api.h`. This include is entirely redundant.

**Fix**: Remove the line entirely (glib.h is already available via transitive includes). If kept, use `#include <glib.h>`.

### 5. Empty `#ifdef GDK_WINDOWING_QUARTZ` block (LOW)

`src/libs/lib.c:35-36`: After removing `#include "osx/osx.h"`, the conditional block is empty. Dead code.

**Fix**: Remove the empty `#ifdef`/`#endif`.

### 6. Widget created unconditionally for all expandable modules (LOW)

`src/libs/lib.c:1344-1349`: `preset_label` widget is created in `dt_lib_gui_get_expander` for **all** expandable modules, even those where `has_preset_label()` defaults to `FALSE`. The widget exists but is never updated (a dead GtkLabel in the layout). Minor memory/waste but also forces the per-frame SQL query described in issue #2 (since `module->preset_label` is non-NULL).

**Fix**: Guard widget creation with `if(module->has_preset_label(module))`.

zisoft commented:
Ok, will look :-)
